build/v1/types.go

```diff
@@ -1022,23 +1026,48 @@ type ImageChangeTrigger struct {
+
+	// name is the name of the ImageStreamTag used for an ImageChangeTrigger
+	Name string `json:"name,omitempty" protobuf:"bytes,2,opt,name=name"`
+}
+
 // ImageChangeTriggerStatus tracks the latest resolved status of the associated ImageChangeTrigger policy
 // specified in the BuildConfigSpec.Triggers struct.
 type ImageChangeTriggerStatus struct {
-	// lastTriggeredImageID is the sha/id of the imageref cited in the 'from' field the last time this BuildConfig was triggered.
-	// It is not necessarily the sha/id of the image change that triggered a build.
-	// This field is updated for all image change triggers when any of them triggers a build.
+	// lastTriggeredImageID represents, at the last time a Build for this BuildConfig was instantiated,  the sha/id of
+	// image referenced by the the ImageStreamTag cited in the 'from' field of a corresponding
+	// ImageChangeTrigger captured by the non-nil BuildTriggerPolicy.ImageChange pointer from a BuildConfigSpec.Triggers
+	// entry.   The lastTriggeredImageID field will be updated regardless of why the last Build was instantiated for this
+	// BuildConfig.
+	// Specifically, an instantiation of a Build could have been manually requested, or could have resulted from
+	// any of the Triggers defined in BuildConfigSpec.Triggers.  For each of those reasons, this field will be updated
+	// with the sha/id of the image referenced by the ImageStreamTag at the time of the Build instantiation.
+	// This field is updated for all image change triggers when any of them triggers a Build to prevent
+	// multiple builds being instantiated when multiple ImageChangeTriggers fire concurrently.
+	// Lastly, to clarify exactly what is meant by "Build instantiation", from a REST perspective, it is a HTTP POST of a
+	// BuildRequest object as the HTTP Body that is made to the OpenShift APIServer, where that HTTP POST also specifies
+	// the "buildconfigs" resource,  "instantiate" subresource, as well as the namespace and name of the BuildConfig.
 	LastTriggeredImageID string `json:"lastTriggeredImageID,omitempty" protobuf:"bytes,1,opt,name=lastTriggeredImageID"`
 
 	// from is the ImageStreamTag that is used as the source of the trigger.
 	// This can come from an ImageStream tag referenced in this BuildConfig's triggers, or the From image in this BuildConfig's build strategy.
-	From *corev1.ObjectReference `json:"from,omitempty" protobuf:"bytes,2,opt,name=from"`
-
-	// paused is true if this trigger is temporarily disabled, and the setting on the spec has been reconciled. Optional.
-	Paused bool `json:"paused,omitempty" protobuf:"varint,3,opt,name=paused"`
-
-	// lastTriggerTime is the last time the BuildConfig was triggered by a change in the ImageStreamTag associated with this trigger.
-	LastTriggerTime metav1.Time `json:"lastTriggerTime,omitempty" protobuf:"bytes,4,opt,name=lastTriggerTime"`
+	From *ImageStreamTagReference `json:"from,omitempty" protobuf:"bytes,2,opt,name=from"`
```

Comment:
Under what conditions is this information not known, but we were able to determine a SHA?  A SHA without telling me which imagestreamtagreference doesn't seem very helpful to general consumers.  It seems like we'd want to optimize for general consumers and if we own code that special cases the buildconfigfrom, it could actually know the value of the imagestream tag.

If for some reason taht isn't practical, indicating what it means when this is nil and how to actually find an imagestreamtag for it seems helpful.

Reply:
Yeah this should not be a pointer.  Obvious in hindsight.  I'll update in next push.  Thanks.

Reply:
to clarify a bit further, the current under the covers behavior where if the spec is nil, they fill in the strategy's from, as the godoc here cites, will be explicit here, meaning the strategy from values will exist in the corresponding entry for it in our new array in the bc.status for ICTs